File: backend/app/models/mediaplan.py
"""MediaPlan — monthly budget breakdown per project."""
import uuid

from sqlalchemy import String, Float, Integer, JSON, ForeignKey, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from datetime import datetime

from app.db.base import Base


class MediaPlan(Base):
    """One record per project — stores monthly plan as JSON rows."""

    __tablename__ = "media_plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False, unique=True
    )
    # JSON list of MonthEntry objects:
    # [{"month": 1, "year": 2025, "pct": 8.0, "budget": 40000,
    #   "forecast_clicks": 1200, "forecast_leads": 60, "cpa": 667}]
    rows: Mapped[list | None] = mapped_column(JSON, nullable=True)
    updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/src/api/mediaplan.ts
import { api } from './client'

export interface MediaPlanRow {
  month: number
  month_name: string
  year: number
  pct: number
  budget: number
  forecast_clicks: number | null
  forecast_leads: number | null
  cpa: number | null
  cpc?: number | null
}

export interface MediaPlan {
  plan_id: string
  rows: MediaPlanRow[]
  total_budget: number
  total_clicks: number
  total_leads: number
  total_frequency: number
  updated_at: string | null
}

export const mediaplanApi = {
  get: (projectId: string) =>
    api.get<MediaPlan>(`/projects/${projectId}/direct/mediaplan`).then((r) => r.data),

  update: (projectId: string, rows: MediaPlanRow[]) =>
    api.put(`/projects/${projectId}/direct/mediaplan`, { rows }).then((r) => r.data),

  reset: (projectId: string, year?: number) =>
    api.post(`/projects/${projectId}/direct/mediaplan/reset`, null, { params: { year } }).then((r) => r.data),
}
